x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -267,7 +270,8 @@ static Set<String> fieldNames(LogicalPlan parsed) {
         references.removeIf(a -> a instanceof MetadataAttribute || MetadataAttribute.isSupported(a.qualifiedName()));
         Set<String> fieldNames = references.names();
         if (fieldNames.isEmpty()) {
-            return IndexResolver.ALL_FIELDS;
+            // there cannot be an empty list of fields, we'll ask the simplest and lightest one instead: _index
+            return IndexResolver.INDEX_METADATA_FIELD;
```

Comment:
So, there are three situations here:
- just a regular list of fields extracted from the plan. Like `from test | keep first_name`
- all fields, simply because they are needed: `from test`
- and no fields at all because we don't need to read anything from Lucene `from test | stats count(*)`
`fieldNames()` job is to decide what `_field_caps` will use in its call.
On one hand  having all this logic in this static method helps greatly with unit testing.
On the other hand, there is logic outside `fieldNames` that changes its content already, like fields needed for enrich policies.

I'll have another look at this. For unit testing's sake, I think I will integrate the enrich policies default match fields logic inside `fieldNames` and add some more tests to `IndexResolverFieldNamesTests`.

Reply:
To clarify, this minor remark is meant as highly optional. Nice-to-have if `fieldnames`'s responsibility vs. `preAnalyzeIndices` would be super clear, but not worth putting in real effort IMO :)

Reply:
I've reworked this. Mostly because of how enrich policies were handled, and it turned out there was a small improvement there. Thanks for raising the discussion point. Even though I haven't addressed your exact concern, something was definitely improved there.